torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py

```diff
@@ -1341,7 +1336,7 @@ def _rebuild_full_params(self,
       dependency_tensors = []
 
     if apply_opt_barrier:
-      self._apply_opt_barrier_to_params_and_tensors(self.full_params,
+      self._apply_opt_barrier_to_params_and_tensors([p for p in self.full_params if p._has_full_param],
```

Comment:
@ronghanghu Wonder why we want to `_apply_opt_barrier_to_params_and_tensors`  to p even if it doesn't hold the full param?

Reply:
@alanwaketan Hmm, it probably isn't necessary to run `_apply_opt_barrier_to_params_and_tensors` on those tensors without full parameters. Back in April 2022, I had some struggle getting the optimization barrier to work. I ended up adding this to a lot of places to finally get it working, but with hindsight it's probably not needed in several places. (The main purpose of this optimization barrier is to prevent the XLA compiler from de-duplicating the two `_rebuild_full_params` in forward and backward passes and holding the full parameters after forward -- which saves collective ops but causes OOM.)